skbio/io/format/genbank.py

```diff
@@ -0,0 +1,468 @@
+from skbio.io.registry import FileSentinel
+from skbio.io.format._base import (_get_nth_sequence,
+                                   _parse_fasta_like_header,
+                                   _format_fasta_like_records, _line_generator,
+                                   _too_many_blanks)
+from skbio.util._misc import chunk_str
+from skbio.sequence import Sequence, DNA, RNA, Protein
+from pprint import pprint
+
+genbank = create_format('genbank')
+
+# date format in locus line of genbank record
+_TIME_FORMAT = '%d-%b-%Y'
+# This list is ordered
+# used to read and write genbank file.
+_HEADERS = ['LOCUS',
+            'DEFINITION',
+            'ACCESSION',
+            'VERSION',
+            'DBSOURCE',
+            'DBLINK',
+            'KEYWORDS',
+            'SOURCE',
+            'REFERENCE',
+            'COMMENT',
+            'FEATURES',
+            'ORIGIN']
+
+
+@genbank.sniffer()
+def _genbank_sniffer(fh):
+    # check the 1st line starts with 'LOCUS' and the last line ends with '//'
+    if _too_many_blanks(fh, 5):
+        return False, {}
+    try:
+        line = next(_line_generator(fh, skip_blanks=True))
+    except StopIteration:
+        return False, {}
+
+    if line.startswith(_HEADERS[0]):
+        for line in _line_generator(fh, skip_blanks=True):
+            if line == '//':
+                return True, {}
+        return False, {}
+    else:
+        return False, {}
+
+
+@genbank.reader(None)
+def _genbank_to_generator(fh, constructor=None, **kwargs):
+    for record in _parse_genbanks(fh):
+        yield _construct(record, constructor, **kwargs)
+
+
+@genbank.reader(Sequence)
+def _genbank_to_biological_sequence(fh, rec_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), rec_num)
+    return _construct(record, Sequence, **kwargs)
+
+
+@genbank.reader(DNA)
+def _genbank_to_dna(fh, rec_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), rec_num)
+    return _construct(record, DNA, **kwargs)
+
+
+@genbank.reader(RNA)
+def _genbank_to_rna(fh, rec_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), rec_num)
+    return _construct(record, DNA, **kwargs).transcribe()
+
+
+@genbank.reader(Protein)
+def _genbank_to_protein(fh, rec_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), rec_num)
+    return _construct(record, Protein, **kwargs)
+
+
+@genbank.writer(None)
+def _sequences_to_genbank(fh):
+    pass
+
+
+def _construct(record, constructor=None, **kwargs):
+    seq, md, pmd = record
+    if constructor is None:
+        unit = md['LOCUS']['unit']
+        if unit == 'bp':
+            # RNA mol type has T instead of U;
+            # so still read in as DNA
+            constructor_ = DNA
+        elif unit == 'aa':
+            constructor_ = Protein
+        else:
+            constructor_ = Sequence
+    else:
+        constructor_ = constructor
+    return constructor_(seq, metadata=md, positional_metadata=pmd, **kwargs)
```

Comment:
sure. since `DNA`, `RNA` and `Protein` all have `lowercase`, I don't see a
reason why `Sequence` shouldn't have the same API of `lowercase`.

On Tue, Aug 11, 2015 at 1:31 PM, Evan Bolyen <notifications@github.com>
wrote:

> In skbio/io/format/genbank.py
> <https://github.com/biocore/scikit-bio/pull/1023#discussion_r36794759>:
>
> > +
> > +def _construct(record, constructor=None, **kwargs):
> > +    seq, md, pmd = record
> > +    if constructor is None:
> > +        unit = md['LOCUS']['unit']
> > +        if unit == 'bp':
> > +            # RNA mol type has T instead of U;
> > +            # so still read in as DNA
> > +            constructor_ = DNA
> > +        elif unit == 'aa':
> > +            constructor_ = Protein
> > +        else:
> > +            constructor_ = Sequence
> > +    else:
> > +        constructor_ = constructor
> > +    return constructor_(seq, metadata=md, positional_metadata=pmd, **kwargs)
>
> Sequence has no alphabet, and so lowercase wasn't needed in the first
> place. I suppose there isn't a strong reason why casing couldn't be handled
> identically for Sequence. For the moment, just do a partial on the needed
> constructors.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1023/files#r36794759>.
>